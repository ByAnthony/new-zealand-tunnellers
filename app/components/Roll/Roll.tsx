"use client";

import { useState } from "react";
import { RollAlphabet } from "./RollAlphabet/RollAlphabet";
import { Title } from "../Title/Title";
import { Tunneller } from "../../types/tunnellers";

import STYLES from "./Roll.module.scss";

type Props = {
  tunnellers: Record<string, Tunneller[]>;
};

export function Roll({ tunnellers }: Props) {
  const [filterByLetter, setFilterByLetter] = useState("");

  if (tunnellers) {
    const letters = Object.keys(tunnellers);
    return (
      <>
        <div className={STYLES.container}>
          <div className={STYLES.header}>
            <Title title={"The New Zealand\\Tunnellers"} />
          </div>
          <div className={STYLES["roll-container"]}>
            <div className={STYLES.alphabet}>
              {letters.map((letter) => (
                <button
                  type="button"
                  key={letter}
                  className={STYLES.letter}
                  onClick={() => setFilterByLetter(letter)}
                  aria-label={`Filter names by the letter ${letter}`}
                >
                  {letter}
                </button>
              ))}
              <button
                type="button"
                key="All"
                className={STYLES.letter}
                onClick={() => setFilterByLetter("")}
                aria-label="Remove the filter by name"
              >
                All
              </button>
            </div>
            <RollAlphabet
              tunnellers={tunnellers}
              filterByLetter={filterByLetter}
            />
          </div>
        </div>
      </>
    );
  }
  return null;
}
