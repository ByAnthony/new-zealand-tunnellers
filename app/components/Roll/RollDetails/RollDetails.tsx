"use client";

import { Tunneller } from "../../../types/tunnellers";
import { displayBiographyDates } from "../../../utils/components/displayBiographyDates";

import STYLES from "./RollDetails.module.scss";

type Props = {
  listOfTunnellers: Tunneller[];
};

export function RollDetails({ listOfTunnellers }: Props) {
  return (
    <>
      {listOfTunnellers.map((tunneller: any) => (
        <a href={`/tunnellers/${tunneller.id}`} key={tunneller.id}>
          <div className={STYLES.tunneller}>
            <div>
              <p className={STYLES.forename}>{tunneller.name.forename}</p>
              <p className={STYLES.surname}>{tunneller.name.surname}</p>
              <p className={STYLES.dates}>
                {displayBiographyDates(
                  tunneller.birthYear,
                  tunneller.deathYear,
                )}
              </p>
            </div>
            <div className={STYLES.arrow}>&rarr;</div>
          </div>
        </a>
      ))}
    </>
  );
}
