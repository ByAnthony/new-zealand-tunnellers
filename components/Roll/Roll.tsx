"use client";

import isEqual from "lodash/isEqual";
import { useCallback, useEffect, useMemo, useRef, useState } from "react";

import { RollAlphabet } from "@/components/Roll/RollAlphabet/RollAlphabet";
import { RollFilter } from "@/components/Roll/RollFilter/RollFilter";
import { RollNoResults } from "@/components/Roll/RollNoResults/RollNoResults";
import { Title } from "@/components/Title/Title";
import { Tunneller } from "@/types/tunnellers";
import { useWindowDimensions } from "@/utils/helpers/useWindowDimensions";

import STYLES from "./Roll.module.scss";
import { getUniqueCorps } from "./utils/corpsUtils";
import { getUniqueDetachments } from "./utils/detachmentUtils";
import {
  getSortedRanks,
  getUniqueRanks,
  rankCategories,
} from "./utils/rankUtils";
import { getUniqueBirthYears, getUniqueDeathYears } from "./utils/yearsUtils";
import { Dialog } from "../Dialog/Dialog";

type Props = {
  tunnellers: Record<string, Tunneller[]>;
};

type Filters = {
  detachment: string[];
  corps: string[];
  ranks: Record<string, string[]>;
  birthYear: string[];
  unknownBirthYear: string;
  deathYear: string[];
  unknownDeathYear: string;
};

export function Roll({ tunnellers }: Props) {
  const { width } = useWindowDimensions();

  /** ---- Derived data ---- */
  const tunnellersList = useMemo(
    () => Object.entries(tunnellers),
    [tunnellers],
  );
  const uniqueDetachments = useMemo(
    () => getUniqueDetachments(tunnellersList),
    [tunnellersList],
  );
  const uniqueCorps = useMemo(
    () => getUniqueCorps(tunnellersList),
    [tunnellersList],
  );
  const uniqueRanks = useMemo(
    () => getUniqueRanks(tunnellersList),
    [tunnellersList],
  );
  const sortedRanks = useMemo(() => getSortedRanks(uniqueRanks), [uniqueRanks]);
  const uniqueBirthYears = useMemo(
    () => getUniqueBirthYears(tunnellersList),
    [tunnellersList],
  );
  const uniqueDeathYears = useMemo(
    () => getUniqueDeathYears(tunnellersList),
    [tunnellersList],
  );

  /** ---- Defaults ---- */
  const defaultFilters: Filters = useMemo(
    () => ({
      detachment: [],
      corps: [],
      ranks: {
        Officers: [],
        "Non-Commissioned Officers": [],
        "Other Ranks": [],
      },
      birthYear: uniqueBirthYears,
      unknownBirthYear: "unknown",
      deathYear: uniqueDeathYears,
      unknownDeathYear: "unknown",
    }),
    [uniqueBirthYears, uniqueDeathYears],
  );

  /** ---- Initial state ---- */
  const [filters, setFilters] = useState<Filters>(defaultFilters);
  const [currentPage, setCurrentPage] = useState<number>(1);

  /** ---- Hydration + persistence gating (no extra render) ---- */
  const hydratedRef = useRef(false);

  /** ---- Read from localStorage ---- */
  const [ready, setReady] = useState(false);

  useEffect(() => {
    if (typeof window === "undefined") return;
    let cancelled = false;

    const storedFilters = localStorage.getItem("filters");
    const storedPage = localStorage.getItem("page");

    Promise.resolve().then(() => {
      if (cancelled) return;

      if (storedFilters) {
        setFilters(JSON.parse(storedFilters));
      }
      if (storedPage) {
        const n = Number(storedPage);
        if (Number.isFinite(n) && n > 0) setCurrentPage(n);
      }

      hydratedRef.current = true;
      setReady(true);
    });

    return () => {
      cancelled = true;
    };
  }, []);

  /** ---- Persist to localStorage AFTER hydration ---- */
  useEffect(() => {
    if (typeof window === "undefined" || !hydratedRef.current) return;

    localStorage.setItem("filters", JSON.stringify(filters));
  }, [filters]);

  useEffect(() => {
    if (typeof window === "undefined" || !hydratedRef.current) return;

    localStorage.setItem("page", String(currentPage));
  }, [currentPage]);

  /** ---- Helpers ---- */
  const hasAnyActiveFilter = (f: Filters): boolean => {
    const ranksActive = Object.values(f.ranks ?? {}).some((arr) => arr.length);
    const detachmentActive = (f.detachment ?? []).length > 0;
    const corpsActive = (f.corps ?? []).length > 0;
    const birthActive =
      (f.birthYear ?? []).length > 0 || f.unknownBirthYear === "unknown";
    const deathActive =
      (f.deathYear ?? []).length > 0 || f.unknownDeathYear === "unknown";
    return (
      ranksActive ||
      detachmentActive ||
      corpsActive ||
      birthActive ||
      deathActive
    );
  };

  /** ---- Filtering ---- */
  const filteredGroups: [string, Tunneller[]][] = useMemo(() => {
    if (!hasAnyActiveFilter(filters)) return [];
    return tunnellersList
      .map<[string, Tunneller[]]>(([group, list]) => [
        group,
        list
          .filter(
            (t) =>
              !filters.detachment?.length ||
              filters.detachment.includes(t.detachment),
          )
          .filter((t) => {
            if (!filters.corps?.length) return true;
            if (
              filters.corps.includes("Tunnelling Corps") &&
              t.attachedCorps === null
            )
              return true;
            return filters.corps.includes(t.attachedCorps ?? "");
          })
          .filter((t) => {
            const r = filters.ranks;
            if (!r || Object.values(r).every((arr) => arr.length === 0))
              return true;
            return Object.values(r).some((arr) => arr.includes(t.rank));
          })
          .filter((t) => {
            const wantsUnknown = filters.unknownBirthYear === "unknown";
            const list = filters.birthYear ?? [];
            if (wantsUnknown && t.birthYear === null) return true;
            if (list.length && t.birthYear) return list.includes(t.birthYear);
            return !wantsUnknown && list.length === 0;
          })
          .filter((t) => {
            const wantsUnknown = filters.unknownDeathYear === "unknown";
            const list = filters.deathYear ?? [];
            if (wantsUnknown && t.deathYear === null) return true;
            if (list.length && t.deathYear) return list.includes(t.deathYear);
            return !wantsUnknown && list.length === 0;
          }),
      ])
      .filter(([, list]) => list.length > 0);
  }, [filters, tunnellersList]);

  const totalFilteredTunnellers = useMemo(
    () => filteredGroups.reduce((acc, [, list]) => acc + list.length, 0),
    [filteredGroups],
  );
  const totalTunnellers = useMemo(
    () => tunnellersList.reduce((acc, [, list]) => acc + list.length, 0),
    [tunnellersList],
  );

  const startBirthYear = filters.birthYear?.[0];
  const endBirthYear = filters.birthYear?.[filters.birthYear.length - 1];
  const startDeathYear = filters.deathYear?.[0];
  const endDeathYear = filters.deathYear?.[filters.deathYear.length - 1];

  /** ---- Handlers ---- */
  const setPageToFirst = useCallback(() => setCurrentPage(1), []);

  const handleDetachmentFilter = useCallback(
    (detachment: string) => {
      setFilters((prev) => {
        const det = new Set(prev.detachment ?? []);
        det.has(detachment) ? det.delete(detachment) : det.add(detachment);
        return { ...prev, detachment: Array.from(det) };
      });
      setPageToFirst();
    },
    [setPageToFirst],
  );

  const handleCorpsFilter = useCallback(
    (corps: string) => {
      setFilters((prev) => {
        const set = new Set(prev.corps ?? []);
        set.has(corps) ? set.delete(corps) : set.add(corps);
        return { ...prev, corps: Array.from(set) };
      });
      setPageToFirst();
    },
    [setPageToFirst],
  );

  const handleBirthSliderChange = useCallback(
    (value: number | number[]) => {
      if (!Array.isArray(value)) return;
      const [start, end] = value;
      setFilters((prev) => ({
        ...prev,
        birthYear: uniqueBirthYears.filter(
          (y) => y >= String(start) && y <= String(end),
        ),
      }));
      setPageToFirst();
    },
    [uniqueBirthYears, setPageToFirst],
  );

  const handleUnknwonBirthYear = useCallback(
    (unknown: string) => {
      setFilters((prev) => ({
        ...prev,
        unknownBirthYear: unknown ? "unknown" : "",
      }));
      setPageToFirst();
    },
    [setPageToFirst],
  );

  const handleDeathSliderChange = useCallback(
    (value: number | number[]) => {
      if (!Array.isArray(value)) return;
      const [start, end] = value;
      setFilters((prev) => ({
        ...prev,
        deathYear: uniqueDeathYears.filter(
          (y) => y >= String(start) && y <= String(end),
        ),
      }));
      setPageToFirst();
    },
    [uniqueDeathYears, setPageToFirst],
  );

  const handleUnknwonDeathYear = useCallback(
    (unknown: string) => {
      setFilters((prev) => ({
        ...prev,
        unknownDeathYear: unknown ? "unknown" : "",
      }));
      setPageToFirst();
    },
    [setPageToFirst],
  );

  const handleRankFilter = useCallback(
    (ranksFilter: Record<string, string[]>) => {
      setFilters((prev) => {
        const nextRanks: Record<string, string[]> = { ...prev.ranks };
        Object.entries(ranksFilter).forEach(([category, ranks]) => {
          if (ranks.length === 0) {
            const allSelected = rankCategories[category].every((rank) =>
              (nextRanks[category] ?? []).includes(rank),
            );
            nextRanks[category] = allSelected ? [] : rankCategories[category];
          } else {
            const set = new Set(nextRanks[category] ?? []);
            ranks.forEach((rank) =>
              set.has(rank) ? set.delete(rank) : set.add(rank),
            );
            nextRanks[category] = Array.from(set);
          }
        });
        return { ...prev, ranks: nextRanks };
      });
      setPageToFirst();
    },
    [setPageToFirst],
  );

  const handleResetFilters = useCallback(() => {
    if (!isEqual(filters, defaultFilters)) {
      setCurrentPage(1);
      setFilters(defaultFilters);
    }
  }, [filters, defaultFilters]);

  /** ---- Dialog state ---- */
  const [isOpen, setIsOpen] = useState(false);
  const onClose = useCallback(() => setIsOpen(false), []);
  const handleFilterButton = useCallback(() => setIsOpen(true), []);

  const rollFiltersProps = {
    className: STYLES["filters-container"],
    uniqueDetachments,
    uniquecorps: uniqueCorps,
    uniqueBirthYears,
    uniqueDeathYears,
    sortedRanks,
    filters,
    startBirthYear,
    endBirthYear,
    startDeathYear,
    endDeathYear,
    handleDetachmentFilter,
    handleCorpsFilter,
    handleBirthSliderChange,
    handleDeathSliderChange,
    handleRankFilter,
    handleUnknwonBirthYear,
    handleUnknwonDeathYear,
  };

  const isDesktop = () => (width ? width > 896 : false);

  return (
    <>
      <Dialog
        id="filter-dialog"
        isFooterEnabled={true}
        isOpen={isOpen}
        handleResetFilters={handleResetFilters}
        onClose={onClose}
        title="Filters"
        totalFiltered={totalFilteredTunnellers}
        total={totalTunnellers}
      >
        <RollFilter {...rollFiltersProps} />
      </Dialog>
      <div className={STYLES.container}>
        <div className={STYLES.header}>
          <Title title={"The New Zealand\\Tunnellers"} />
        </div>
        {ready && (
          <div className={STYLES["roll-container"]}>
            <div className={STYLES.controls}>
              <div className={STYLES["results-container"]}>
                <button
                  className={STYLES["reset-button"]}
                  onClick={handleResetFilters}
                >
                  Reset filters
                </button>
                <p className={STYLES.results}>
                  {`${totalFilteredTunnellers} result${totalFilteredTunnellers > 1 ? "s" : ""}`}
                </p>
              </div>
              <button
                className={STYLES["filter-button"]}
                onClick={handleFilterButton}
              >
                Filters
              </button>
              {isDesktop() ? <RollFilter {...rollFiltersProps} /> : null}
            </div>
            {filteredGroups.length > 0 ? (
              <RollAlphabet
                tunnellers={filteredGroups}
                currentPage={currentPage}
                onPageChange={setCurrentPage}
              />
            ) : (
              <RollNoResults handleResetFilters={handleResetFilters} />
            )}
          </div>
        )}
      </div>
    </>
  );
}
