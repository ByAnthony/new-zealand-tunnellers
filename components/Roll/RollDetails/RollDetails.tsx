"use client";

import Link from "next/link";

import { Tunneller } from "@/types/tunnellers";
import { displayBiographyDates } from "@/utils/helpers/roll";

import STYLES from "./RollDetails.module.scss";

type Props = {
  listOfTunnellers: Tunneller[];
};

export function RollDetails({ listOfTunnellers }: Props) {
  return (
    <>
      {listOfTunnellers.map((tunneller: Tunneller) => (
        <Link href={`/tunnellers/${tunneller.id}`} key={tunneller.id}>
          <div className={STYLES.tunneller}>
            <div>
              <p className={STYLES.rank}>{tunneller.rank}</p>
              <p className={STYLES.forename}>{tunneller.name.forename}</p>
              <p className={STYLES.surname}>{tunneller.name.surname}</p>
              <p className={STYLES.detachment}>{tunneller.detachment}</p>
              <p className={STYLES.dates}>
                {displayBiographyDates(
                  tunneller.birthYear,
                  tunneller.deathYear,
                )}
              </p>
            </div>
            <div className={STYLES.arrow}>&rarr;</div>
          </div>
        </Link>
      ))}
    </>
  );
}
