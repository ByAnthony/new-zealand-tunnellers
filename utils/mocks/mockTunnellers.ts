import { Tunneller, TunnellerWithFullNameData } from "@/types/tunnellers";

const mockTunnellerData1: TunnellerWithFullNameData = {
  id: 1,
  forename: "John",
  surname: "Doe",
  birthYear: "1886",
  deathYear: "1952",
  fullName: "John Doe",
};

const mockTunnellerData2: TunnellerWithFullNameData = {
  id: 2,
  forename: "Biff",
  surname: "Tanen",
  birthYear: "1897",
  deathYear: null,
  fullName: "William Smith",
};

const mockTunnellerData3: TunnellerWithFullNameData = {
  id: 3,
  forename: "Emmett",
  surname: "Brown",
  birthYear: null,
  deathYear: "1935",
  fullName: "Emmett Brown",
};

const mockTunnellerData4: TunnellerWithFullNameData = {
  id: 4,
  forename: "Marty",
  surname: "McFly",
  birthYear: null,
  deathYear: null,
  fullName: "Marty McFly",
};

export const mockTunnellersData: TunnellerWithFullNameData[] = [
  mockTunnellerData1,
  mockTunnellerData2,
  mockTunnellerData3,
  mockTunnellerData4,
];

const mockTunneller1: Tunneller = {
  id: 1,
  name: { forename: "John", surname: "Doe" },
  birthYear: "1886",
  deathYear: "1952",
  fullName: "John Doe",
};

const mockTunneller2: Tunneller = {
  id: 2,
  name: { forename: "Biff", surname: "Tanen" },
  birthYear: "1897",
  deathYear: null,
  fullName: "William Smith",
};

const mockTunneller3: Tunneller = {
  id: 3,
  name: { forename: "Emmett", surname: "Brown" },
  birthYear: null,
  deathYear: "1935",
  fullName: "Emmett Brown",
};

const mockTunneller4: Tunneller = {
  id: 4,
  name: { forename: "Marty", surname: "McFly" },
  birthYear: null,
  deathYear: null,
  fullName: "Marty McFly",
};

export const mockTunnellers: Record<string, Tunneller[]> = {
  B: [mockTunneller3],
  D: [mockTunneller1],
  M: [mockTunneller4],
  S: [mockTunneller2],
};
