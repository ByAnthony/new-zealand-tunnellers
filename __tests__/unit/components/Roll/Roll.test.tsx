import { fireEvent, render, screen } from "@testing-library/react";

import { Roll } from "@/components/Roll/Roll";
import { AttachedCorpsBadge } from "@/components/Roll/RollDetails/RollDetails";
import { mockTunnellers } from "__tests__/unit/utils/mocks/mockTunnellers";

describe("Roll", () => {
  beforeEach(() => {
    global.scrollTo = jest.fn();
  });

  test("matches the snapshot", () => {
    const { asFragment } = render(<Roll tunnellers={mockTunnellers} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("renders the component correctly", () => {
    render(<Roll tunnellers={mockTunnellers} />);

    const buttonB = screen.getByLabelText("Filter names by the letter B");
    const buttonD = screen.getByLabelText("Filter names by the letter D");
    const buttonM = screen.getByLabelText("Filter names by the letter M");
    const buttonT = screen.getByLabelText("Filter names by the letter T");
    expect(buttonB).toBeInTheDocument();
    expect(buttonB).toHaveTextContent("B");
    expect(buttonD).toBeInTheDocument();
    expect(buttonD).toHaveTextContent("D");
    expect(buttonM).toBeInTheDocument();
    expect(buttonM).toHaveTextContent("M");
    expect(buttonT).toBeInTheDocument();
    expect(buttonT).toHaveTextContent("T");

    const buttonAll = screen.getByLabelText("Remove the filter by letter");
    expect(buttonAll).toBeInTheDocument();
    expect(buttonAll).toHaveTextContent("All");

    const titleB = screen.getByLabelText("Letter B");
    const titleD = screen.getByLabelText("Letter D");
    const titleM = screen.getByLabelText("Letter M");
    const titleT = screen.getByLabelText("Letter T");
    expect(titleB).toHaveClass("title");
    expect(titleB).toHaveTextContent("B");
    expect(titleB).toBeInTheDocument();
    expect(titleD).toHaveClass("title");
    expect(titleD).toHaveTextContent("D");
    expect(titleD).toBeInTheDocument();
    expect(titleM).toHaveClass("title");
    expect(titleM).toHaveTextContent("M");
    expect(titleM).toBeInTheDocument();
    expect(titleT).toHaveClass("title");
    expect(titleT).toHaveTextContent("T");
    expect(titleT).toBeInTheDocument();

    expect(screen.getByText("John")).toBeInTheDocument();
    expect(screen.getByText("Doe")).toBeInTheDocument();
    const johnDoe = screen.getByRole("link", {
      name: "Sapper John Doe Main Body 1886-1952 →",
    });
    expect(johnDoe).toHaveAttribute("href", "/tunnellers/1");

    expect(screen.getByText("Biff")).toBeInTheDocument();
    expect(screen.getByText("Tanen")).toBeInTheDocument();
    const biffTanen = screen.getByRole("link", {
      name: "Sapper Biff Tanen 2nd Reinforcements 1897-†? →",
    });
    expect(biffTanen).toHaveAttribute("href", "/tunnellers/2");

    expect(screen.getByText("Emmett")).toBeInTheDocument();
    expect(screen.getByText("Brown")).toBeInTheDocument();
    const EmmetBrown = screen.getByRole("link", {
      name: "Sapper Emmett Brown Main Body ?-1935 →",
    });
    expect(EmmetBrown).toHaveAttribute("href", "/tunnellers/3");

    expect(screen.getByText("Marty")).toBeInTheDocument();
    expect(screen.getByText("McFly")).toBeInTheDocument();
    const MartyMcFly = screen.getByRole("link", {
      name: "Sapper Marty McFly 5th Reinforcements Army Pay Corps ?-†? →",
    });
    expect(MartyMcFly).toHaveAttribute("href", "/tunnellers/4");
  });

  test("should filter by name", () => {
    render(<Roll tunnellers={mockTunnellers} />);

    const buttonD = screen.getByLabelText("Filter names by the letter D");
    fireEvent.click(buttonD);
    expect(screen.getByLabelText("Letter D")).toBeInTheDocument();
    expect(screen.queryByLabelText("Letter B")).not.toBeInTheDocument();
    expect(screen.queryByLabelText("Letter M")).not.toBeInTheDocument();
    expect(screen.queryByLabelText("Letter S")).not.toBeInTheDocument();

    expect(screen.getByText("John")).toBeInTheDocument();
    expect(screen.getByText("Doe")).toBeInTheDocument();

    expect(screen.queryByText("Biff")).not.toBeInTheDocument();
    expect(screen.queryByText("Tanen")).not.toBeInTheDocument();

    expect(screen.queryByText("Emmett")).not.toBeInTheDocument();
    expect(screen.queryByText("Brown")).not.toBeInTheDocument();

    expect(screen.queryByText("Marty")).not.toBeInTheDocument();
    expect(screen.queryByText("McFly")).not.toBeInTheDocument();
  });

  test("should remove filter by name", () => {
    render(<Roll tunnellers={mockTunnellers} />);

    const buttonD = screen.getByLabelText("Filter names by the letter D");
    const buttonAll = screen.getByLabelText("Remove the filter by letter");

    fireEvent.click(buttonD);

    expect(screen.getByLabelText("Letter D")).toBeInTheDocument();
    expect(screen.queryByLabelText("Letter B")).not.toBeInTheDocument();
    expect(screen.queryByLabelText("Letter M")).not.toBeInTheDocument();
    expect(screen.queryByLabelText("Letter T")).not.toBeInTheDocument();

    fireEvent.click(buttonAll);

    expect(screen.getByLabelText("Letter D")).toBeInTheDocument();
    expect(screen.getByLabelText("Letter B")).toBeInTheDocument();
    expect(screen.getByLabelText("Letter M")).toBeInTheDocument();
    expect(screen.getByLabelText("Letter T")).toBeInTheDocument();
  });
});

describe("AttachedCorpsBadge", () => {
  test("renders correctly with given props", () => {
    render(<AttachedCorpsBadge attachedCorps="Engineers" />);
    const badgeElement = screen.getByText("Engineers");
    expect(badgeElement).toBeInTheDocument();
  });
});
