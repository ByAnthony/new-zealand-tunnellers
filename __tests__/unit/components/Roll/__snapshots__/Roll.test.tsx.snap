// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Roll matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="header"
    >
      <div
        class="main-title"
      >
        <h1>
          <span
            class="title-line-1"
          >
            The New Zealand
          </span>
          <span
            class="title-line-2"
          >
            Tunnellers
          </span>
        </h1>
      </div>
    </div>
    <div
      class="roll-container"
    >
      <div
        class="controls"
      >
        <div
          class="results-container"
        >
          <button
            class="reset-button"
          >
            Reset filters
          </button>
          <p
            class="results"
          >
            4 results
          </p>
        </div>
        <button
          class="filter-button"
        >
          Filters
        </button>
        <div
          class="filters-container"
        >
          <div
            class="filters"
          >
            <h3>
              Detachments
            </h3>
            <div>
              <label>
                <input
                  id="Main Body"
                  name="Main Body"
                  type="checkbox"
                  value="Main Body"
                />
                Main Body
              </label>
            </div>
            <div>
              <label>
                <input
                  id="2nd Reinforcements"
                  name="2nd Reinforcements"
                  type="checkbox"
                  value="2nd Reinforcements"
                />
                2nd Reinforcements
              </label>
            </div>
            <div>
              <label>
                <input
                  id="5th Reinforcements"
                  name="5th Reinforcements"
                  type="checkbox"
                  value="5th Reinforcements"
                />
                5th Reinforcements
              </label>
            </div>
          </div>
          <div
            class="filters"
          >
            <h3>
              Corps
            </h3>
            <div>
              <label>
                <input
                  id="Tunnelling Corps"
                  name="Tunnelling Corps"
                  type="checkbox"
                  value="Tunnelling Corps"
                />
                Tunnelling Corps
              </label>
            </div>
            <div>
              <label>
                <input
                  id="Army Pay Corps"
                  name="Army Pay Corps"
                  type="checkbox"
                  value="Army Pay Corps"
                />
                Army Pay Corps
              </label>
            </div>
          </div>
          <div
            class="filters"
          >
            <h3>
              Birth Years
            </h3>
            <p>
              1886-1897
            </p>
            <div
              class="slider"
            >
              <div
                class="rc-slider rc-slider-horizontal"
              >
                <div
                  class="rc-slider-rail"
                  style="background: rgb(64, 66, 67);"
                />
                <div
                  class="rc-slider-track rc-slider-track-1"
                  style="left: 0%; width: 100%; background: rgb(153, 131, 100);"
                />
                <div
                  class="rc-slider-step"
                />
                <div
                  aria-disabled="false"
                  aria-orientation="horizontal"
                  aria-valuemax="1897"
                  aria-valuemin="1886"
                  aria-valuenow="1886"
                  class="rc-slider-handle rc-slider-handle-1"
                  role="slider"
                  style="left: 0%; transform: translateX(-50%); border: 2px solid rgb(153, 131, 100); background: rgb(29, 31, 32); outline: none; box-shadow: 0 0 5px rgba(64, 66, 67, 0.5);"
                  tabindex="0"
                />
                <div
                  aria-disabled="false"
                  aria-orientation="horizontal"
                  aria-valuemax="1897"
                  aria-valuemin="1886"
                  aria-valuenow="1897"
                  class="rc-slider-handle rc-slider-handle-2"
                  role="slider"
                  style="left: 100%; transform: translateX(-50%); border: 2px solid rgb(153, 131, 100); background: rgb(29, 31, 32); outline: none; box-shadow: 0 0 5px rgba(64, 66, 67, 0.5);"
                  tabindex="0"
                />
              </div>
            </div>
            <div
              style="margin-top: 15px;"
            >
              <label>
                <input
                  checked=""
                  id="unknownBirthYear"
                  name="unknownBirthYear"
                  type="checkbox"
                  value="unknownBirthYear"
                />
                Includes unknown birth year
              </label>
            </div>
          </div>
          <div
            class="filters"
          >
            <h3>
              Death Years
            </h3>
            <p>
              1935-1952
            </p>
            <div
              class="slider"
            >
              <div
                class="rc-slider rc-slider-horizontal"
              >
                <div
                  class="rc-slider-rail"
                  style="background: rgb(64, 66, 67);"
                />
                <div
                  class="rc-slider-track rc-slider-track-1"
                  style="left: 0%; width: 100%; background: rgb(153, 131, 100);"
                />
                <div
                  class="rc-slider-step"
                />
                <div
                  aria-disabled="false"
                  aria-orientation="horizontal"
                  aria-valuemax="1952"
                  aria-valuemin="1935"
                  aria-valuenow="1935"
                  class="rc-slider-handle rc-slider-handle-1"
                  role="slider"
                  style="left: 0%; transform: translateX(-50%); border: 2px solid rgb(153, 131, 100); background: rgb(29, 31, 32); outline: none; box-shadow: 0 0 5px rgba(64, 66, 67, 0.5);"
                  tabindex="0"
                />
                <div
                  aria-disabled="false"
                  aria-orientation="horizontal"
                  aria-valuemax="1952"
                  aria-valuemin="1935"
                  aria-valuenow="1952"
                  class="rc-slider-handle rc-slider-handle-2"
                  role="slider"
                  style="left: 100%; transform: translateX(-50%); border: 2px solid rgb(153, 131, 100); background: rgb(29, 31, 32); outline: none; box-shadow: 0 0 5px rgba(64, 66, 67, 0.5);"
                  tabindex="0"
                />
              </div>
            </div>
            <div
              style="margin-top: 15px;"
            >
              <label>
                <input
                  checked=""
                  id="unknownDeathYear"
                  name="unknownDeathYear"
                  type="checkbox"
                  value="unknownDeathYear"
                />
                Includes unknown death year
              </label>
            </div>
            <div
              class="filters"
            >
              <h3>
                Ranks
              </h3>
              <div
                style="margin-bottom: 15px;"
              >
                <label
                  class="rank-category"
                >
                  <input
                    id="Other Ranks"
                    name="Other Ranks"
                    type="checkbox"
                    value="Other Ranks"
                  />
                  Other Ranks
                </label>
                <div
                  style="margin-left: 15px;"
                >
                  <label>
                    <input
                      id="Sapper"
                      name="Sapper"
                      type="checkbox"
                      value="Sapper"
                    />
                    Sapper
                  </label>
                </div>
                <div
                  style="margin-left: 15px;"
                >
                  <label>
                    <input
                      id="Driver"
                      name="Driver"
                      type="checkbox"
                      value="Driver"
                    />
                    Driver
                  </label>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="roll"
      >
        <div
          id="letter-B"
        >
          <div
            class="letter-container"
          >
            <h2
              aria-label="Letter B"
              class="title"
            >
              B
            </h2>
          </div>
          <div
            class="tunnellers-container"
          >
            <a
              href="/tunnellers/3"
            >
              <div
                class="tunneller"
              >
                <div>
                  <p
                    class="rank"
                  >
                    Sapper
                  </p>
                  <p
                    class="forename"
                  >
                    Emmett
                  </p>
                  <p
                    class="surname"
                  >
                    Brown
                  </p>
                  <p
                    class="detachment"
                  >
                    Main Body 
                  </p>
                  <p
                    class="dates"
                  >
                    ?-1935
                  </p>
                </div>
                <div
                  class="arrow"
                >
                  →
                </div>
              </div>
            </a>
          </div>
        </div>
        <div
          id="letter-D"
        >
          <div
            class="letter-container"
          >
            <h2
              aria-label="Letter D"
              class="title"
            >
              D
            </h2>
          </div>
          <div
            class="tunnellers-container"
          >
            <a
              href="/tunnellers/1"
            >
              <div
                class="tunneller"
              >
                <div>
                  <p
                    class="rank"
                  >
                    Sapper
                  </p>
                  <p
                    class="forename"
                  >
                    John
                  </p>
                  <p
                    class="surname"
                  >
                    Doe
                  </p>
                  <p
                    class="detachment"
                  >
                    Main Body 
                  </p>
                  <p
                    class="dates"
                  >
                    1886-1952
                  </p>
                </div>
                <div
                  class="arrow"
                >
                  →
                </div>
              </div>
            </a>
          </div>
        </div>
        <div
          id="letter-M"
        >
          <div
            class="letter-container"
          >
            <h2
              aria-label="Letter M"
              class="title"
            >
              M
            </h2>
          </div>
          <div
            class="tunnellers-container"
          >
            <a
              href="/tunnellers/4"
            >
              <div
                class="tunneller"
              >
                <div>
                  <p
                    class="rank"
                  >
                    Driver
                  </p>
                  <p
                    class="forename"
                  >
                    Marty
                  </p>
                  <p
                    class="surname"
                  >
                    McFly
                  </p>
                  <p
                    class="detachment"
                  >
                    5th Reinforcements 
                    <span
                      class="badge"
                    >
                      Army Pay Corps
                    </span>
                  </p>
                  <p
                    class="dates"
                  >
                    ?-†?
                  </p>
                </div>
                <div
                  class="arrow"
                >
                  →
                </div>
              </div>
            </a>
          </div>
        </div>
        <div
          id="letter-T"
        >
          <div
            class="letter-container"
          >
            <h2
              aria-label="Letter T"
              class="title"
            >
              T
            </h2>
          </div>
          <div
            class="tunnellers-container"
          >
            <a
              href="/tunnellers/2"
            >
              <div
                class="tunneller"
              >
                <div>
                  <p
                    class="rank"
                  >
                    Sapper
                  </p>
                  <p
                    class="forename"
                  >
                    Biff
                  </p>
                  <p
                    class="surname"
                  >
                    Tanen
                  </p>
                  <p
                    class="detachment"
                  >
                    2nd Reinforcements 
                  </p>
                  <p
                    class="dates"
                  >
                    1897-†?
                  </p>
                </div>
                <div
                  class="arrow"
                >
                  →
                </div>
              </div>
            </a>
          </div>
        </div>
        <div
          class="pagination"
        >
          <button
            class="pagination-main-button"
            disabled=""
          >
            <span
              class="previous-arrow"
            >
              ‣
            </span>
          </button>
          <button
            class="active"
            disabled=""
          >
            1
          </button>
          <button
            class="pagination-main-button"
            disabled=""
          >
            ‣
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
